### ISSUE TITLE
loseWriteConnection breaks loseConnection

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3037]
|<img alt="mwicked's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| mwicked reported|
|-|-|
|Trac ID|trac#3037|
|Type|defect|
|Created|2008-02-15 17:39:35Z|
|Branch|https://github.com/twisted/twisted/tree/select-halfclose-3037|

ListenerProtocol1 leaks socket fd's each time it answers a query, and connectionLost is never triggered.  ListenerProtocol2 works as expected.

```python
class ListenerProtocol1(Protocol):
    """ Protocol that knows how to deal with connections to our listener """

    implements(IHalfCloseableProtocol)

    def __init__(self, poller):
        self.poller = poller
        self.data = []

    def dataReceived(self, data):
        self.data.append(data)

    def readConnectionLost(self):
        query = "".join(self.data).lower()
        self.transport.loseWriteConnection()

    def writeConnectionLost(self):
        self.transport.write("YO")
        log.msg("finished responding to query")
        self.transport.loseWriteConnection()

    def connectionLost(self, reason):
        log.msg("finished responding to query")




class ListenerProtocol2(Protocol):
    """ Protocol that knows how to deal with connections to our listener """

    implements(IHalfCloseableProtocol)

    def __init__(self, poller):
        self.poller = poller
        self.data = []

    def dataReceived(self, data):
        self.data.append(data)

    def readConnectionLost(self):
        query = "".join(self.data).lower()
        self.transport.write("Yo")
        self.transport.loseConnection()

    def connectionLost(self, reason):
        log.msg("finished responding to query")
```

Attachments:

* [3037.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/c5a/c5aa7a64b76b97100a08380d254f46559e72c05d/226610182bc90fc1763dbdd447840ed6e009f253.patch) (1804 bytes) - added by zubin71 on 2010-09-29 13:43:54Z - All the changes mentioned above have been made.

<details><summary>Searchable metadata</summary>

```
trac-id__3037 3037
type__defect defect
reporter__mwicked mwicked
priority__normal normal
milestone__ 
branch__branches_select_halfclose_3037 branches/select-halfclose-3037
branch_author__itamarst itamarst
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1203097175000000 1203097175000000
changetime__1321231796000000 1321231796000000
version__None None
owner__itamar itamar
cc__jesstess
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

Editing markup.

**Comment by twisted-trac** [comment]
|<img alt="mwicked's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_2"></a>|mwicked commented|
|-|-|

Sorry, there is a typo in the problem.  In the ListenerProtocol1, the second call to loseWriteConnection should be loseConnection instead.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_3"></a>|@glyph set owner to @jyknight|
|-|-|

Reassigning since you are the guy who likes half-close.

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)<a name="note_4"></a>|@jyknight commented|
|-|-|

Test case:
```
from twisted.internet import protocol, interfaces
from zope.interface import implements
class ListenerProtocol1(protocol.Protocol):
    """ Protocol that knows how to deal with connections to our listener """

    implements(interfaces.IHalfCloseableProtocol)

    def __init__(self):
        self.data = []

    def dataReceived(self, data):
        print "dataReceived %r" % data
        self.data.append(data)

    def readConnectionLost(self):
        print "readConnectionLost"
        self.transport.loseWriteConnection()

    def writeConnectionLost(self):
        print "writeConnectionLost"
        self.transport.loseConnection()

    def connectionLost(self, reason):
        print "connectionLost"


class ListenerFactory(protocol.Factory):
    protocol = ListenerProtocol1


#from twisted.internet import epollreactor
#epollreactor.install()

from twisted.internet import reactor
port = reactor.listenTCP(5555, ListenerFactory())
reactor.run()
```

Then run
```
python -c 'import time, socket; s=socket.socket(); 
s.connect(("localhost", 5555)); s.shutdown(1); time.sleep(5)'
```

Appears to only exhibit in the select reactor on Linux, not in epoll, poll, not select on OSX.

This is because Linux 'select' stops indicating writeability for the fd after you call shutdown(SHUT_WR).

However, twisted already contains the workaround for this, in ```twisted/internet/abstract.py loseConnection```: if ```self._writeDisconnected``` is true, it disconnects immediately, instead of waiting for writeability. That is all that's necessary here...

This doesn't properly trigger, however, because in doWrite(), it sets ```_writeDisconnected = True``` *after* ```self._closeWriteConnection()``` instead of *before*. I think swapping those two statements will fix the bug.

Should be easy to add the test case to the test suite and verify that that change doesn't break anything else, so adding easy keyword. :)

**Comment by twisted-trac** [comment]
|<img alt="zubin71's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_5"></a>|zubin71 removed owner|
|-|-|

The patch to be reviewed is 3037.patch . 
A new test twisted/internet/test/test_abstract.py has been added.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun set owner to igorsobreira|
|-|-|

The test needs to be re-written to be a unit test - subclass `ReactorBuilder` or add the methods to an existing class in `twisted/internet/test/test_tcp.py`.  As it is written now, the test will corrupt global state and hang indefinitely (at best).

assigning to igorsobreira because he expressed some interested on IRC in working on this.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_7"></a>|@itamarst commented|
|-|-|

(In [33104]) Branching to 'select-halfclose-3037'

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_8"></a>|@itamarst removed owner|
|-|-|

I couldn't reproduce the error on my machine, but it did happen on a bunch of the buildslaves; the proposed fix did indeed work.

http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/select-halfclose-3037

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)|@jesstess set owner to @jesstess|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_10"></a>|@jesstess set owner to @itamarst|
|-|-|

I talked with Itamar about this after being unable to reproduce on Linux as well, and he said all the Windows buildslaves and some of the Linux ones would fail on the test case.

 * "I was previously asked to to" ==> "I was previously asked to"

Other than that, looks good, please merge.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_11"></a>|@itamarst set status to `closed`|
|-|-|

(In [33142]) Merge select-halfclose-3037.

Author: mwicked, jknight, itamar
Review: jesstess
Fixes: [#3037](https://github.com/twisted/twisted/issues/3037)

Fix edge-case where notification of half-close never happens.
